tasks/human_eval_018.rs

```diff
@@ -9,10 +9,144 @@ use vstd::prelude::*;
+    ensures
+        (substring[0] == string.subrange(0, substring.len() as int)[0] && (substring.skip(1)
+            =~= string.skip(1).subrange(0, substring.skip(1).len() as int))) ==> (substring
+            =~= string.subrange(0, substring.len() as int)),
+    decreases substring.len(),
+{
+    if (substring[0] == string.subrange(0, substring.len() as int)[0] && (substring.skip(1)
+        =~= string.skip(1).subrange(0, substring.skip(1).len() as int))) {
+        assert forall|i: int| 0 <= i < substring.len() implies #[trigger] substring[i]
+            == string.subrange(0, substring.len() as int)[i] by {
+            if i == 0 {
+            } else {
+                assert(forall|j: int|
+                    (0 <= #[trigger] (j + 0) < substring.len() - 1) ==> substring.skip(1)[j]
+                        == string.skip(1).subrange(0, substring.skip(1).len() as int)[j]);
+                assert(0 <= #[trigger] (i - 1 + 0) < substring.len() - 1);
+            }
+        }
+    } else {
+    }
+}
+
+fn is_prefix(substring: Vec<char>, string: Vec<char>) -> (b: bool)
+    ensures
+        b == substring@.is_prefix_of(string@),
+{
+    let mut current_substring = substring.clone();
+    let mut current_string = string.clone();
+
+    if substring.len() > string.len() {
+        return false
+    }
+    while (current_substring.len() > 0)
+        invariant
+            0 <= current_substring.len() <= current_string.len(),
+            substring.len() <= string.len(),
+            (substring@ =~= string@.subrange(0, substring@.len() as int)) == (current_substring@
+                =~= current_string@.subrange(0, current_substring@.len() as int)),
+        decreases current_substring.len(),
+    {
+        if (current_substring[0] != current_string[0]) {
+            return false;
+        }
+        let old_substring = current_substring.clone();
+        let old_string = current_string.clone();
+
+        let substring_first = current_substring.remove(0);
+        let string_first = current_string.remove(0);
+        assert((old_substring@ =~= old_string@.subrange(0, old_substring@.len() as int)) <== (
+        old_substring@[0] == old_string@.subrange(0, old_substring@.len() as int)[0] && (
+        old_substring@.skip(1) =~= old_string@.skip(1).subrange(
+            0,
+            old_substring@.skip(1).len() as int,
+        )))) by { lemma_step_subrange(old_substring@, old_string@) };
+    }
+    return true;
+}
+
+proof fn lemma_how_many_times_zero(string: Seq<char>, substring: Seq<char>)
+    requires
+        string.len() < substring.len(),
+    ensures
+        how_many_times(string, substring) == 0,
+    decreases string.len(),
+{
+    if string.len() == 0 {
+    } else {
+        lemma_how_many_times_zero(string.skip(1), substring)
+    }
+}
+
+fn how_many_times_impl(string: Vec<char>, substring: Vec<char>) -> (opt_k: Option<u32>)
+    requires
+        substring.len() >= 1,
+    ensures
+        match opt_k {
```

Comment:
I think it would also be reasonable to require `string.len() <= u32::MAX` instead of returning an `Option`.  But if you like this version, feel free to leave as is.

Reply:
Hmm I think I slightly prefer the way it is now

Reply:
Okay, I'll merge it in as is then.